docs/standard/io/pipelines.md

```diff
@@ -0,0 +1,343 @@
+
+The following code reads all messages from a `PipeReader` and calls `ProcessMessageAsync` on each.
+
+[!code-csharp[MyConnection1](~/samples/snippets/csharp/pipelines/MyConnection1.cs?name=snippet)]
+
+### Cancellation
+
+`PipeReader.ReadAsync`:
+
+* Supports passing a <xref:System.Threading.CancellationToken>.
+* Throws in an <xref:System.OperationCanceledException> if the `CancellationToken` is canceled while there's a read pending.
+* Supports a way to cancel the current read operation via <xref:System.IO.Pipelines.PipeReader.CancelPendingRead%2A?displayProperty=nameWithType>, which avoids raising an exception. Calling `PipeReader.CancelPendingRead` returns a <xref:System.IO.Pipelines.ReadResult> with `IsCanceled` set to `true`. This can be useful for halting the existing read loop in a non-destructive and non-exceptional way.
+
+[!code-csharp[MyConnection](~/samples/snippets/csharp/pipelines/MyConnection.cs?name=snippet)]
+
+<a name="gotchas"></a>
+
+### PipeReader common problems
+
+* Passing the wrong values to `consumed` or `examined` may result in reading already read data. For example, passing a `SequencePosition` that was already processed.
+* Passing `buffer.End` as examined may result in:
+
+  * Stalled data
+  * Possibly an eventual Out of Memory (OOM) exception if data isn't consumed. For example, `PipeReader.AdvanceTo(position, buffer.End)` when processing a single message at a time from the buffer.
+
+* Passing the wrong values to `consumed` or `examined` may result in an infinite loop. For example, `PipeReader.AdvanceTo(buffer.Start)` if `buffer.Start` hasn't changed will cause the next call to `PipeReader.ReadAsync` to return immediately before new data arrives.
+* Passing the wrong values to `consumed` or `examined` may result in infinite buffering (eventual OOM). For example, `PipeReader.AdvanceTo(buffer.Start, buffer.End)` unconditionally when processing a single message at a time from the buffer.
+* Using the `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo` may result in memory corruption (use after free).
+* Failing to call `PipeReader.Complete/CompleteAsync` may result in a memory leak.
+* Checking <xref:System.IO.Pipelines.ReadResult.IsCompleted?displayProperty=nameWithType> and exiting the reading logic before processing the buffer results in data loss. The loop exit condition should be based on `ReadResult.Buffer.IsEmpty` and `ReadResult.IsCompleted`. Doing this in the wrong order could result in an infinite loop.
+
+#### Problematic code
+
+❌ **Data loss**
+
+The `ReadResult` can return the final segment of data when `IsCompleted` is set to `true`. Not reading that data before exiting the read loop will result in data loss.
+
+[!INCLUDE [pipelines-do-not-use-1](../../../includes/pipelines-do-not-use-1.md)]
+
+[!code-csharp[DoNotUse#1](~/samples/snippets/csharp/pipelines/DoNotUse.cs?name=snippet)]
+
+[!INCLUDE [pipelines-do-not-use-2](../../../includes/pipelines-do-not-use-2.md)]
```

Comment:
do we really need a before and after note?

Reply:
re: bad code includes before/after
> do we really need a before and after note?

@mairaw  for bad code examples, my vote is yes. People google on API's and copy/paste the first hit sample they find.

BEtter would be to remove the copy button AND put a big red X through the code and disable copy.

Reply:
Ok, I posted on the support channel about this. Hopefully, they'll create a customer suggestion based on that.